File: ball3Dtrajectory.py
#This code calculates a 3D trajectory of an arbitrarily spinning ball
#subject to gravity, aerodynamic drag, and the Magnus force
import numpy as np
import math
import pandas as pd
#drag coefficient from v and S according to Clanet 2015
def cdfunc(vx,vy,vz,w,r,vc,vs):
    v=math.sqrt(vx**2+vy**2+vz**2);
    S=w*r/v;
    if S>0.05 and v>vc:
        cd=0.4127*S**0.3056;
    else:
        cd=0.155+0.346/(1+math.exp((v-vc)/vs))
    return cd
#Function for the numeric integration
def numericintegration(vx0,vy0,vz0,x0,y0,z0,wx,wy,wz,vc,vs,r,m,cl):
    kt=0
    #Total angular frequency
    w=math.sqrt(wx**2+wy**2+wz**2)
    #time step size (vary to check the convergence)
    dto1=1/220;
    #Initiating position, time and velocity
    xo1=np.array([x0]);
    yo1=np.array([y0]);
    zo1=np.array([z0]);
    vxo1=np.array([vx0]);
    vyo1=np.array([vy0]);
    vzo1=np.array([vz0]);
    to1=np.array([0]);
    #step counter
    ko1=1;
    #First order numeric integration
    while min(yo1)>-0.001:  #a little less then zero to take out the initial point with z=0.
        cd=cdfunc(vxo1[-1],vyo1[-1],vzo1[-1],w,r,vc,vs)
        vo1=math.sqrt(vxo1[-1]**2+vyo1[-1]**2+vzo1[-1]**2)
        axo1=-1/(2*m)*cd*A*rho*vo1**2*vxo1[-1]/vo1\
        +cl/m*math.pi*r**3*rho*(wy*vzo1[-1]-wz*vyo1[-1]);
        ayo1=-1/(2*m)*cd*A*rho*vo1**2*vyo1[-1]/vo1-g\
        +cl/m*math.pi*r**3*rho*(wz*vxo1[-1]-wx*vzo1[-1]);
        azo1=-1/(2*m)*cd*A*rho*vo1**2*vzo1[-1]/vo1\
            +cl/m*math.pi*r**3*rho*(wx*vyo1[-1]-wy*vxo1[-1]);
        vxo1=np.append(vxo1,np.array([vxo1[-1]+axo1*dto1]));
        vyo1=np.append(vyo1,np.array([vyo1[-1]+ayo1*dto1]));
        vzo1=np.append(vzo1,np.array([vzo1[-1]+azo1*dto1]));
        xo1=np.append(xo1,np.array([xo1[-1]+vxo1[-1]*dto1]));
        yo1=np.append(yo1,np.array([yo1[-1]+vyo1[-1]*dto1]));
        zo1=np.append(zo1,np.array([zo1[-1]+vzo1[-1]*dto1]));
        to1=np.append(to1,np.array([to1[-1]+dto1]));
        
        ko1=ko1+1;


    data=np.array([to1,xo1,yo1,zo1,vxo1,vyo1,vzo1])
    
    return data
#Constants
g=9.81;
rho=1.2; 

#ball
r=0.11;    #radius
A=math.pi*r**2; #projected or frontal area of the ball [m^2]
m=0.430;         #kg


#default drag and lift coefficients 
cd=0.15;         #default ball drag coefficient
cl=1.7/math.pi;       #lift coefficient 


#For the CD-model
vc=12.19; 
vs=1.309;
#Starting position
x0=0;
y0=0;
z0=0;
#Starting angular velocity
wx=0;
wy=2*math.pi*13.5;
wz=0;
w=math.sqrt(wx**2+wy**2+wz**2)
#Roberto Carlos 1997 :13.5 rps, 38 m/s
#initial velocity
v0=40;                    #initial velocity in the (x,y) plane
theta0=45;                #initial angle to the horizontal
vz0=0;                    #initial velocity component in z direction
vtot0=math.sqrt(v0**2+vz0**2);   #initial total speed of the ball
vx0=v0*math.cos(theta0*math.pi/180) #In cartesian coordinates
vy0=v0*math.sin(theta0*math.pi/180)
#non-dimensional parameters
Ap=v0**2/(r*g);                    #range parameter A
B=m*g/(0.5*cd*A*rho*v0**2);   #ballistic parameter B
S=w*r/vtot0;                     #spin parameter S


Thang=2*(vy0/g)
tvec=np.linspace(0,Thang,2200)
#Run the function with the given data (You can make several initial conditions...
#and compare them. Just make sure to change the input and give the output another name eg.)
#data2=numericintegration(vx02,vy02,vz02,x02,y02,z02,wx2,wy2,wz2,vc,vs,r,m,cl)
#Remember to define the new variables first.
data=numericintegration(vx0,vy0,vz0,x0,y0,z0,wx,wy,wz,vc,vs,r,m,cl)
#Extract the x,y,z and t-values
t = data[0,:]
x = data[1,:]
y = data[2,:]
z = data[3,:]
rangexo1 = max(data[1,:]);
rangezo1 = max(data[2,:]);
heighto1 = max(data[3,:]);
hangtimeo1=max(data[0,:]);
import matplotlib.pyplot as plt # Import the matplotlib.pyplot module
#Compare vacuum and numeric air solution
plt.figure(10)
plt.subplot(2,1,1);
plt.tick_params(axis='both',bottom=True, top=True, left=True, right=True, direction='in', which='major')
plt.axis('equal')
plt.box
plt.plot(x, y, "b-",linewidth=4)
#Analytic
plt.plot(vx0*tvec, vy0*tvec-1/2*g*(tvec)**2, "g-",linewidth=2)
#Measured is plotted twice to make sure the data can be viewed
plt.xlabel("x [m]") # Set the x-axis label
plt.ylabel("y [m]") # Set the y-axis label
plt.legend(['Num,air','Ana,vac'],loc='best',ncol=3)
plt.subplot(2,1,2);
plt.tick_params(axis='both',bottom=True, top=True, left=True, right=True, direction='in', which='major')
plt.axis('equal')
plt.box
plt.plot(x, z, "b-",linewidth=4)
#Analytic
plt.plot(vx0*tvec, vz0*tvec, "g-",linewidth=2)
#Measured is plotted twice to make sure the data can be viewed
plt.xlabel("x [m]") # Set the x-axis label
plt.ylabel("z [m]") # Set the y-axis label
plt.legend(['Num,air','Ana,vac'],loc='best',ncol=3)
plt.show()

File: ball3Dtrajectory.m
%This code calculates a 3D trajectory of an arbitrarily spinning ball
%subject to gravity, aerodynamic drag, and the Magnus force

clear; close all;

%physical parameters
g=9.81;       %[m/s^2]
rho=1.2;       % air density [kg/m^3]

%ball parameters
r=0.11;        %football radius [m]
Aproj=pi*r^2;  %projected or frontal area of the ball [m^2]
m=0.430;       %football kg

%default drag and lift coefficients 
cd=0.15;         %default ball drag coefficient
cl=1.7/pi;       %lift coefficient 

%initial position
x0=0;
y0=0;
z0=0;

%Roberto Carlos 1997 :13.5 rps, 38 m/s
%initial velocity
v0=38;                    %initial velocity in the (x,y) plane
theta0=45;                %initial angle to the horizontal
vz0=0;                    %initial velocity component in z direction
vtot0=sqrt(v0^2+vz0^2);   %initial total speed of the ball

%the spin vector is assumed to be constant during the flight
wx=0;
wy=2*pi*13.5;   
wz=0;

%magnitude of spin vector
w=sqrt(wx^2+wy^2+wz^2);


%non-dimensional parameters
A=v0^2/(r*g);                    %range parameter A
B=m*g/(0.5*cd*Aproj*rho*v0^2);   %ballistic parameter B
S=w*r/vtot0;                     %spin parameter S

%critical velocity for calculation of drag coefficient CD
vc=12.19; 
vs=1.309;

% here different loops are performed, which can be replaced by other loops
% such as %for m=[0.001 0.01 0.1 1 10], %for theta0=38:1:4, %for
% dto1=logspace(0,-3,4), or %for wy=[-10:2:10]*2*pi

%loop over different trajectories
%trajectory counter
kt=0;

%for v0=10:10:40
for v0=38 
kt=kt+1;
    
    %calculate velocity components
    vx0=cosd(theta0)*v0;
    vy0=sind(theta0)*v0;
    
    
    %characteristic parameters of the vacuum solution 
    hangtime_vac(kt)=2*vy0/g;
    height_vac(kt)=vy0^2/(2*g);
    range_vac(kt)=v0^2*sind(2*theta0)/g;
    
    %calculate analytic vacuum trajectory
    tvac(:,kt) = linspace(0,hangtime_vac(kt));
    xvac(:,kt) = vx0*tvac(:,kt);
    yvac(:,kt) = y0+vy0*tvac(:,kt)-g/2*tvac(:,kt).^2;
    
    
    %first-order numerical integration 
   
    %time step size (vary to check the convergence)
    dto1=0.001; %s
    
    %initial conditions
    xo1(1,kt)=x0;
    yo1(1,kt)=y0;
    zo1(1,kt)=z0;
    vxo1(1,kt)=vx0;
    vyo1(1,kt)=vy0;
    vzo1(1,kt)=vz0;
    to1(1,kt)=0;
    
    %step counter
    ko1=1;
    while min(yo1(:,kt))> -0.001  %a little less then zero to take out the initial point with z=0.
        
        %drag coefficient from v and S
        vo1(ko1,kt)=sqrt(vxo1(ko1,kt)^2+vyo1(ko1,kt)^2+vzo1(ko1,kt)^2);
        S=w*r/vo1(ko1,kt);
        if S>0.05 && vo1(ko1,kt)>vc
            cd=0.4127*S^0.3056;
            cdo1(ko1,kt)=0.4127*S^0.3056;
        else
            cd=0.155+0.346/(1+exp((vo1(ko1,kt)-vc)/vs));
            cdo1(ko1,kt)=0.155+0.346/(1+exp((vo1(ko1,kt)-vc)/vs));
        end
        
        %accelerations due to gravity, aerodyanmic drag and the Magnus force 
        axo1=-1/(2*m)*cd*Aproj*rho*vo1(ko1,kt)^2*vxo1(ko1,kt)/vo1(ko1,kt)...
            +cl/m*pi*r^3*rho*(wy*vzo1(ko1,kt)-wz*vyo1(ko1,kt));
        ayo1=-1/(2*m)*cd*Aproj*rho*vo1(ko1,kt)^2*vyo1(ko1,kt)/vo1(ko1,kt)-g...
            +cl/m*pi*r^3*rho*(wz*vxo1(ko1,kt)-wx*vzo1(ko1,kt));
        azo1=-1/(2*m)*cd*Aproj*rho*vo1(ko1,kt)^2*vzo1(ko1,kt)/vo1(ko1,kt)...
            +cl/m*pi*r^3*rho*(wx*vyo1(ko1,kt)-wy*vxo1(ko1,kt));
        
        %first-order integration of the acceleration gives the velocity,
        %and another first-order integration gives the position
        vxo1(ko1+1,kt)=vxo1(ko1,kt)+axo1*dto1;
        vyo1(ko1+1,kt)=vyo1(ko1,kt)+ayo1*dto1;
        vzo1(ko1+1,kt)=vzo1(ko1,kt)+azo1*dto1;
        xo1(ko1+1,kt)=xo1(ko1,kt)+vxo1(ko1,kt)*dto1;
        yo1(ko1+1,kt)=yo1(ko1,kt)+vyo1(ko1,kt)*dto1;
        zo1(ko1+1,kt)=zo1(ko1,kt)+vzo1(ko1,kt)*dto1;
        
        %advance the time
        to1(ko1+1,kt)=to1(ko1,kt)+dto1;
        
        ko1=ko1+1;
    end
    
    %numerically calculated characteristic trajectory parameters
    rangexo1(kt) = max(xo1(:,kt));
    rangezo1(kt) = max(zo1(:,kt));
    heighto1(kt) = max(yo1(:,kt));
    hangtimeo1(kt)=max(to1(:,kt));
    
    
    %this section uses the Runge-Kutta scheme implemented in Matlab for the trajectory calculation    
    tspan=[0 1.1*hangtimeo1(kt)];
    eq1=@(t,x)[x(2);1/m*(-1/2*cdfunc(x(2),x(4),x(6),w,r,vc,vs)*Aproj*rho*(x(2)^2+x(4)^2+x(6)^2)*x(2)/(sqrt(x(2)^2+x(4)^2+x(6)^2))...
        +cl*pi*r^3*rho*(wy*x(6)-wz*x(4)));...
        x(4);1/m*(-m*g-1/2*cdfunc(x(2),x(4),x(6),w,r,vc,vs)*Aproj*rho*(x(2)^2+x(4)^2+x(6)^2)*x(4)/(sqrt(x(2)^2+x(4)^2+x(6)^2))...
        +cl*pi*r^3*rho*(wz*x(2)-wx*x(6)));...
        x(6);1/m*(-1/2*cdfunc(x(2),x(4),x(6),w,r,vc,vs)*Aproj*rho*(x(2)^2+x(4)^2+x(6)^2)*x(6)/(sqrt(x(2)^2+x(4)^2+x(6)^2))...
        +cl*pi*r^3*rho*(wx*x(4)-wy*x(2)))];
    
    sol1=ode45(eq1,tspan,[x0 vx0 y0 vy0 z0 vz0]);
    trk45 = linspace(0,1.1*hangtimeo1(kt));
    xrk45(:,kt) = deval(sol1,trk45,1);
    vxrk45(:,kt)= deval(sol1,trk45,2);
    yrk45(:,kt) = deval(sol1,trk45,3);
    vyrk45(:,kt)= deval(sol1,trk45,4);
    zrk45(:,kt) = deval(sol1,trk45,5);
    vzrk45(:,kt)= deval(sol1,trk45,6);
    
    xrk45(find(yrk45(:,kt)<-eps),kt)=nan;
    vxrk45(find(yrk45(:,kt)<-eps),kt)=nan;
    vyrk45(find(yrk45(:,kt)<-eps),kt)=nan;
    zrk45(find(yrk45(:,kt)<-eps),kt)=nan;
    vzrk45(find(yrk45(:,kt)<-eps),kt)=nan;
    yrk45(find(yrk45(:,kt)<-eps),kt)=nan;
end

%the num1 solution has different lengths for different trajectories,
%we replace last zeros with nan to prevent plotting

[npoints,ntrajectories]=size(xo1);
for kp=2:npoints
    for kt=1:ntrajectories
        if xo1(kp,kt)<eps
            xo1(kp,kt)=nan;
            yo1(kp,kt)=nan;
            zo1(kp,kt)=nan;
        end
    end
end



%compare vacuum, the order 1, and the Runge-Kutta solutions
figure(10); clf; 
subplot(2,1,1); hold on; box on;
plot(xrk45,yrk45)
plot(xo1,yo1)
plot(xvac,yvac)
xlabel('x [m]')
ylabel('y [m]')
xlim([0 1.1*max(max(max(xvac)),max(max(xo1)))])
ylim([0 1.1*max(max(max(yvac)),max(max(yo1)))])
title('x-y')
hold off
subplot(2,1,2); hold on; box on;
plot(xrk45,zrk45)
plot(xo1,zo1)
xlabel('x [m]')
ylabel('z [m]')
axis equal
title('x-z')
hold off


% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% %drag coefficient function for Runge-Kutta integration
function cd=cdfunc(vx,vy,vz,w,radius,vc,vs)
v=sqrt(vx^2+vy^2+vz^2);
S=w*radius/v;
        if S>0.05 && v>vc
            cd=0.4127*S^0.3056;
        else
            cd=0.155+0.346/(1+exp((v-vc)/vs));
        end

end
